src/Visualization/Force/QuadTree.elm

```diff
@@ -0,0 +1,317 @@
+        , children = ( vertex, [] )
+        , aggregate = ()
+        }
+
+
+size : QuadTree aggregate vertex -> Int
+size qtree =
+    case qtree of
+        Empty ->
+            0
+
+        Leaf leaf ->
+            let
+                ( _, rest ) =
+                    leaf.children
+            in
+                1 + List.length rest
+
+        Node node ->
+            size node.nw + size node.ne + size node.se + size node.sw
+
+
+insertBy : (vertex -> Point2d) -> vertex -> QuadTree () vertex -> QuadTree () vertex
+insertBy toPoint vertex qtree =
+    case qtree of
+        Empty ->
+            Leaf
+                { boundingBox = BoundingBox2d.singleton (toPoint vertex)
+                , children = ( vertex, [] )
+                , aggregate = ()
+                }
+
+        Leaf leaf ->
+            let
+                ( first, rest ) =
+                    leaf.children
+
+                newSize =
+                    2 + List.length rest
+
+                -- this seems to be a good choice.
+                -- normally, all vertices are put into one large leaf. When it reaches this size
+                -- the leaf is split. From then on the node will resize when it encounters a point outside of its bounding box.
+                -- resizing is expensive, so we wait a while before splitting. On the other hand, if we never split at all,
+                -- we get none of the benefits of using an quadtree.
+                maxSize =
+                    32
+            in
+                if newSize >= maxSize then
+                    let
+                        initial =
+                            { ne = [], nw = [], se = [], sw = [] }
+
+                        folder element accum =
+                            case quadrant leaf.boundingBox (toPoint element) of
+                                NE ->
+                                    { accum | ne = element :: accum.ne }
+
+                                SE ->
+                                    { accum | se = element :: accum.se }
+
+                                NW ->
+                                    { accum | nw = element :: accum.nw }
+
+                                SW ->
+                                    { accum | sw = element :: accum.sw }
+
+                        byQuadrant : { ne : List vertex, nw : List vertex, se : List vertex, sw : List vertex }
+                        byQuadrant =
+                            List.foldl folder initial (vertex :: first :: rest)
+                    in
+                        Node
+                            { boundingBox = BoundingBox2d.hull leaf.boundingBox (BoundingBox2d.singleton (toPoint vertex))
+                            , ne = fromList toPoint byQuadrant.ne
+                            , se = fromList toPoint byQuadrant.se
+                            , nw = fromList toPoint byQuadrant.nw
+                            , sw = fromList toPoint byQuadrant.sw
```

Comment:
Does it make sense to create the intermediate lists here, or should you just call `insertBy` directly from the `folder`? I would assume that would save on allocating the list without really any loss in clarity. I may be wrong about that...

Reply:
In fact, that would make the code identical to the beginning of the Node branch...

Reply:
using a fold is exactly as fast (so no speed gain) but way shorter, thanks!